### PR TITLE
TUP-697: User Updates, Change Page Listing to Use a Plugin

### PR DESCRIPTION
**Pull request by R-Tomas-Gonzalez**
## Overview

Designers would like to edit the User Updates page. Created a plugin that adds in all of the user updates, and then adjusted the page.

## Related

- [TUP-697](https://tacc-main.atlassian.net/browse/TUP-697)

## Changes

1. Makes the default base news/latest-news
2. Adds in full list of User Updates
3. Adds in the template 

## Testing

1. Head over to http://localhost:8000/news/user-updates/
2. Remove text plugin
3. Add in the Full User News List plugin
4. Determine that the page is editable, and you're able to view the articles.

## UI

| Before | After |
| --- | --- |
| Screenshot 2024-03-05 at 11 24 40 PM | Screenshot 2024-03-05 at 11 24 53 PM |

### CODE REVIEW

**Comment by R-Tomas-Gonzalez** [comment]
> **It works!** With some issues.
> 
> 1. Remove "User Updates" from output.
>    So CMS admin can add/remove/change heading without developers.[1](#user-content-fn-1-b244c2d94a6964718839d4c73ec9c491)
> 2. Fix URL for User Update pages.
>    Example: Click the name of a User Update, you get http://localhost:8000/news/latest-news107514/.
> 3. Retire the "User Updates" view.
>    E.g. remove `UserNewsListView`; have `user_news/list.html` or `user_news/full_list.html`, not both; rename `user_news/list_for_plugin.html` (cuz now **all** the plugin content is via plugins).
> 4. Rename plugins. I'm thinking:
>    
>    * "User News List" → "Latest User Updates"[2](#user-content-fn-2-b244c2d94a6964718839d4c73ec9c491)
>    * "Full User News List" → "User Updates"[2](#user-content-fn-2-b244c2d94a6964718839d4c73ec9c491)
> 
> If I've neglected or misunderstood anything, please correct me.
> 
> ## Footnotes
> 1. Consider how we can safely give admins more power. It gives us less work and gets them results faster (cuz to change this text we must change code, test, PR, review, deploy). [↩](#user-content-fnref-1-b244c2d94a6964718839d4c73ec9c491)
> 2. No "List" because they are both a list (and Blog plugin only uses "List" to distinguish) and I don't think there will be other "User Updates" plugins. If there are, then we can append "List". [↩](#user-content-fnref-2-b244c2d94a6964718839d4c73ec9c491) [↩2](#user-content-fnref-2-2-b244c2d94a6964718839d4c73ec9c491)

@wesleyboar 1 and 2 have been addressed by: [6ef537e](https://github.com/TACC/tup-ui/pull/438/commits/6ef537e2c44410b3baa4b3d17518a78cb4c7e94e)

For 3. I have removed `UserNewsListView`. For the rest of this comment, I am confused as to only use one of the files... I guess I'm unsure because I don't know how to use the `list.html` for what we need. The `full_list.html` file shows ALL user updates, and omits the header. The `list.html` has the header included... Is there a way to create an if statement to know what plugin is using the file? Therefore allowing me to only use what is needed from the file that has more?

For 4. Changing names breaks the cms. Need more time to debug that. Will re-review later tonight and will update this comment.

File: apps/tup-cms/src/apps/user_news/cms_plugins.py [hunk 1]
```diff
@@ -32,3 +32,24 @@ def render(self, context, instance, placeholder):
             'urls': urls
         })
         return context
+
+
+@plugin_pool.register_plugin
+class UserUpdatesPlugin(CMSPluginBase):
+    module = 'TUP CMS'
+    name = _('User Updates')
```
Comment (wesleyboar) [inline apps/tup-cms/src/apps/user_news/cms_plugins.py:40]: [Background discussion.](https://github.com/TACC/tup-ui/pull/438#issuecomment-1982034249)

**4. ~~Rename plugins.~~**
> @R-Tomas-Gonzalez: The [User News List] plugin is already being used with it's name and therefore the template is looking for this plugin name. If we remove the plugin, we must delete first before deploying this change. […] We can make this change ("Full User News List" → "User Updates").

Well, crap. Thanks. Okay, then your original name is better. Please revert `e4ced86`.

I will consider whether to make a new ticket to rename user-facing name of both plugins (a semantic courtesy for the CMS admin).[^1]

[^1]: I think https://stackoverflow.com/questions/77105970/how-to-rename-a-djangocmss-plugin and https://gist.github.com/yakky/31466588da712e25e1f2c200d56d3776 could help.

File: apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+<div class="container">
+    <section class="c-news c-news--list">
+      {% include './includes/content_styles.html' %}
+  
+      {% with title_tag="h3" subtitle_tag="h4" %}
+      {% include './includes/list.html' %}
+      {% endwith %}
+  
+    </section>
+</div>
```
Comment (wesleyboar) [inline apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html:12]: **5. Unsuitable markup.** Remove the container (and re-indent afterward). The container here is adding unexpected margin around the list. Such margin is already provided around the page content. [^1]

<details><summary>I.e. this suggestion diff.</summary>

```suggestion
<section class="c-news c-news--list">
  {% include './includes/content_styles.html' %}
  
  {% with title_tag="h3" subtitle_tag="h4" %}
  {% include './includes/list.html' %}
  {% endwith %}
</section>
```

</details>

| Existing Margin (from page Container) | Undesired Margin (from plugin Container) |
| - | - |
| existing |  erroneous |

[^1]: A container is used (on our sites) to wrap **all** content of a page, **not** a single block of content within a page that has multiple blocks. But this container is doing the latter.

File: apps/tup-cms/src/apps/user_news/cms_plugins.py
Comment (wesleyboar) [inline apps/tup-cms/src/apps/user_news/cms_plugins.py:1]: **6. Update Naming.** Given that this app now offers two plugins, not just one, update instances of `is_for_plugin` variable name to be an accurate name, perhaps:
- `is_for_short_list` and _retain_ the values
- `is_for_full_list` and _negate_ the values

**Review by wesleyboar — COMMENTED** [review]
I've broken the PR task discussion into pieces, and requested two new items.

File: apps/tup-cms/src/apps/user_news/views.py [hunk 3]
```diff
@@ -3,19 +3,6 @@
 from .defaults import urls
 from .utils import get_latest_articles, get_article
 
-class UserNewsListView(TemplateView):
-  template_name = 'user_news/list.html'
```
[the remark this answers is not in the document]
Comment (R-Tomas-Gonzalez) [reply apps/tup-cms/src/apps/user_news/views.py:7]: I think you meant, delete `user_news/list.html`. Also, I didn't realize there was a `user_news/list.html` and `includes/list.html` which is why I was confused as to how I was supposed to delete `user_news/list.html`

File: apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html
[the remark this answers is not in the document]
Comment (R-Tomas-Gonzalez) [reply apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html:1]: Done.

File: apps/tup-cms/src/apps/user_news/cms_plugins.py
Comment (R-Tomas-Gonzalez) [reply apps/tup-cms/src/apps/user_news/cms_plugins.py:1]: Fixed.

File: apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html (hunk 2 above)
Comment (R-Tomas-Gonzalez) [reply apps/tup-cms/src/apps/user_news/templates/user_news/full_list.html:12]: Fixed.

File: apps/tup-cms/src/apps/user_news/cms_plugins.py (hunk 1 above)
Comment (R-Tomas-Gonzalez) [reply apps/tup-cms/src/apps/user_news/cms_plugins.py]: Reverted.

File: apps/tup-cms/src/apps/user_news/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wesleyboar) [reply apps/tup-cms/src/apps/user_news/views.py:7]: I did. Correct. Understood.

**Review by wesleyboar — APPROVED** [review]

**Pull request merged**